src/padding.rs

```diff
@@ -1,12 +1,166 @@
+
+fn encapsulate(pkt: &[u8]) -> Result<Vec<u8>, std::io::Error> {
+    // if the packet does not start and end with braces, it is not valid BSON data
+    if pkt[0] != b'{' || pkt[pkt.len() - 1] != b'}' {
+        return Err(Error::from(ErrorKind::InvalidData));
+    }
+
+    let mut rng = rand_chacha::ChaCha8Rng::from_entropy(); // use the ChaCha8Rng algorithm to generate our random numbers and bytes
+    let r1 = generate_random_num(&mut rng); // length of random bytes before the BSON data
+    let r2 = generate_random_num(&mut rng); // length of random bytes after the BSON data
+
+    // vectors containing the random bytes for the front and back padding of the BSON data
+    let front_pad = generate_random_bytes(&mut rng, r1, '{');
+    let back_pad = generate_random_bytes(&mut rng, r2, '}');
+
+    // define a new vector to add the front pad, the packet, and finally the back pad
+    let mut encapsulated_pkt = Vec::new();
+    encapsulated_pkt.extend(front_pad);
+    encapsulated_pkt.extend(pkt);
+    encapsulated_pkt.extend(back_pad);
+
+    Ok(encapsulated_pkt)
 }
 
 /// Strips the padding from an unencrypted packet.
-/// Returns Ok(Vec<u8>) containing the BSON if successful.
-/// Returns Err if the packet is invalid.
+///
+/// # Arguments
+///
+/// * `pkt` - A reference to a BSON packet with randomized padding.
+///
+/// # Returns
+///
+/// The revealed BSON in Vec<u8> form if successful.
+///
+/// # Errors
+///
+/// Returns an error if the BSON packet is not valid (it does not start and end with the correct braces) or
+/// if there is an error during the extraction of BSON data.
 fn reveal(pkt: &[u8]) -> Result<Vec<u8>, std::io::Error> {
-    unimplemented!()
+    let pkt_vec = pkt.to_vec();
+
+    // initialize bson_front_index & bson_back_index with a default of -1 to indicate they hasn't been changed
+    let mut bson_front_index: i32 = -1;
+    let mut bson_back_index = -1;
+
+    // loop through the packet and find the index of the beginning of the BSON data.
+    for (index, &element) in pkt_vec.iter().enumerate() {
+        if element == b'{' {
+            // if the brace is not within 2-32 (inclusive) bytes, the packet is not valid
+            if index < 2 || index > 32 {
+                return Err(Error::from(ErrorKind::InvalidData));
+            } else {
+                bson_front_index = index as i32; // store the index of the bracket for later use
+                break;
+            }
+        }
+    }
+
+    // if the bson_front_index hasn't been changed, we were unable to find an open bracket and therefore need to throw an error
+    if bson_front_index == -1 {
+        return Err(Error::from(ErrorKind::InvalidData));
+    }
+
+    // loop through the packet backwards to find the end of the BSON data
+    for (rev_index, &element) in pkt_vec.iter().rev().enumerate() {
+        if element == b'}' {
+            if rev_index < 2 || rev_index > 32 {
```

Comment:
we should short-circuit after 32 instead of just going on forever. this also applies in the forward direction